### PR TITLE
FOGL-7687 Add fixture to install & add external services in conftest.py

### PR DESCRIPTION
**Pull request by Mohit04tomar**

### CODE REVIEW

File: tests/system/python/conftest.py [hunk 1]
```diff
@@ -212,6 +214,131 @@ def clone_make_install():
 
     return _add_fledge_north
 
+@pytest.fixture
+def add_service():
+    def _add_service(fledge_url, service_branch, installation_type = "make", service_name = "play",
+                     plugin_lang="python", use_pip_cache = True, enabled = True, wait_time, retries):
+        """ Add Service Instance and start the insatnce by default """
+        
+        _enabled = "true" if enabled else "false"
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:223]: API support boolean also. You can use `enabled` as is.

File: tests/system/python/conftest.py [hunk 2]
```diff
@@ -212,6 +214,131 @@ def clone_make_install():
...
+                                   rule_plugin_lang="python", delivery_plugin_lang="python", installation_type='make', notification_type="one shot",
+                                   notification_instance_name="play", retrigger_time=30, use_pip_cache=True, enabled=True):
+        
+        if rule_plugin_discovery_name is None:
+            rule_plugin_discovery_name = rule_plugin
+            
+        if delivery_plugin_discovery_name is None:
+            delivery_plugin_discovery_name = delivery_plugin
+            
+        _enabled = "true" if enabled else "false"
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:289]: Boolean value should also work.

**Review by ashish-jabble — CHANGES_REQUESTED** [review]

File: tests/system/python/conftest.py [hunk 3]
```diff
@@ -212,6 +214,156 @@ def clone_make_install():
...
+    def _add_service(fledge_url, service, service_branch, retries, installation_type = "make", service_name = "play",
+                     use_pip_cache = True, enabled = True):
+        
+        """ 
+            Fixture to add Service and start the start service by default 
+            service: Service to be installed
+            service_branch: Branch of service to be installed
+            retries: Number of tries for polling
+            installation_type: Type of installation for service i.e. make or package
+            serice_name: Name that will be given to service to be installed
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:228]: service_name

File: tests/system/python/conftest.py [hunk 4]
```diff
@@ -212,6 +214,156 @@ def clone_make_install():
...
+            use_pip_cache: use pip cache in requirements installation
+            enabled: Flag to enable or disable notification instace
+        """
+        
+        retval = utils.get_request(fledge_url, "/fledge/service")
+        for ele in retval["services"]:
+            if ele["type"].lower() == service:
+                return ele
+        
+        PROJECT_ROOT = Path(__file__).parent.parent.parent.parent
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:238]: This is a wrong practice to find the parent root.

File: tests/system/python/conftest.py [hunk 5]
```diff
@@ -212,6 +214,156 @@ def clone_make_install():
...
+        if installation_type == 'make':
+            clone_make_install()
+        elif installation_type == 'package':
+            try:
+                subprocess.run(["sudo {} install -y fledge-service-{}".format(pytest.PKG_MGR, service)], shell=True,
+                            check=True)
+            except subprocess.CalledProcessError:
+                assert False, "{} package installation failed!".format(service)
+        else:
+            print("Skipped {} service installation. Installation mechanism is set to {}.".format(service, installation_type))
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:256]: Can't we return from here?

File: tests/system/python/conftest.py [hunk 6]
```diff
@@ -212,6 +214,156 @@ def clone_make_install():
...
+    return _add_service
+
+@pytest.fixture
+def add_notification_instance():
+    def _add_notification_instance(fledge_url, delivery_plugin, delivery_branch , rule_config, delivery_config, rule_plugin="Threshold", 
+                                   rule_branch=None, rule_plugin_discovery_name=None, delivery_plugin_discovery_name=None,
+                                   installation_type='make', notification_type="one shot", notification_instance_name="play", 
+                                   retrigger_time=30, use_pip_cache=True, enabled=True):
+        """
+            Fixture to add Service Instance and start the insatnce by default
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:292]: instance

File: tests/system/python/conftest.py [hunk 7]
```diff
@@ -212,6 +214,156 @@ def clone_make_install():
...
+            delivery_plugin: Notify or Delivery plugin to be installed
+            delivery_branch: Branch of Notify or Delivery plugin to be installed
+            rule_config: Configuration required for Rule plugin
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:295]: it's an optional; if not given then it should we with default config.

File: tests/system/python/conftest.py [hunk 8]
```diff
@@ -212,6 +214,131 @@ def clone_make_install():
...
+        elif service_plugin == "dispatcher":
+            #  Restart fledge and verify dispatcher service is added
+            utils.put_request(fledge_url, "/fledge/restart", None)
+            for i in range(retries):
+                time.sleep(30)
+                get_url = '/fledge/ping'
+                ping_result = utils.get_request(fledge_url, get_url)
+                if ping_result['uptime'] > 0:
+                    break
+            assert ping_result['uptime'] > 0
```
[the remark this answers is not in the document]
Comment (ashish-jabble) [reply tests/system/python/conftest.py:275]: Above way which I mentioned is for both installation and for service registration as well. I don't think it's good idea to restart the fledge and to check dispatcher online or not. We should add service with POST endpoint.

File: tests/system/python/conftest.py (hunk 4 above)
Comment (Mohit04tomar) [reply tests/system/python/conftest.py:238]: I know and I have found the new best way to find the project root i.e. `PROJECT_ROOT = subprocess.getoutput("git rev-parse --show-toplevel")`, but Praveen Sir Suggested me to keep old way of finding the PROJECT_ROOT and in other PR update all occurences of PROJECT_ROOT with `PROJECT_ROOT = subprocess.getoutput("git rev-parse --show-toplevel")`

File: tests/system/python/scripts/install_c_plugin [hunk 9]
```diff
@@ -40,8 +45,9 @@ install_requirement (){
 }
 
 install_binary_file () {
-   if [[ "${PLUGIN_TYPE}" = "rule" ]]
+   if [[ "${PLUGIN_TYPE}" = "rule" || "${PLUGIN_TYPE}" == "notify" ]]
```
[the remark this answers is not in the document]
Comment (Mohit04tomar) [reply tests/system/python/scripts/install_c_plugin:48]: Yes.

File: tests/system/python/conftest.py [hunk 10]
```diff
@@ -313,6 +313,9 @@ def _add_notification_instance(fledge_url, delivery_plugin, delivery_branch , ru
         if delivery_plugin_discovery_name is None:
             delivery_plugin_discovery_name = delivery_plugin
 
+        _rule_config = rule_config if rule_config is not None else {}
+        _delivery_config = delivery_config if delivery_config is not None else {}
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:317]: Can't we set by default to `{}`?
```
_add_notification_instance(....delivery_config={}, rule_config={}...
```

File: tests/system/python/conftest.py [hunk 11]
```diff
@@ -18,6 +18,8 @@
 from urllib.parse import quote
 from pathlib import Path
 import pytest
+from helpers import utils
+import time
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:22]: standard way "import time" should be placed before "import pytest"

File: tests/system/python/conftest.py [hunk 12]
```diff
@@ -212,6 +214,158 @@ def clone_make_install():
...
+        
+        def clone_make_install(plugin_branch, plugin_type, plugin):
+            try:
+                subprocess.run(["{}/tests/system/python/scripts/install_c_plugin {} {} {}".format(
+                    PROJECT_ROOT, plugin_branch, plugin_type, plugin)], shell=True, check=True)
+            except subprocess.CalledProcessError:
+                assert False, "{} plugin installation failed".format(plugin)
+
+        if installation_type == 'make':
+            # First Install Rule Plugin if Rule Plugin is not Threshold
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:327]: This comment is confusing...

File: tests/system/python/conftest.py [hunk 13]
```diff
@@ -212,6 +214,158 @@ def clone_make_install():
...
+                subprocess.run(["sudo {} install -y fledge-service-{}".format(pytest.PKG_MGR, service)], shell=True,
+                            check=True)
+            except subprocess.CalledProcessError:
+                assert False, "{} package installation failed!".format(service)
+        else:
+            return("Skipped {} service installation. Installation mechanism is set to {}.".format(service, installation_type))
+        
+        
+        # Add service
+        if service == "notification":
```
[the remark this answers is not in the document]
Comment (Mohit04tomar) [reply tests/system/python/conftest.py:260]: In fledge repository, we only keep system test of plugins and services that are in fledge-iot organization.

[the remark this answers is not in the document]
Comment (ashish-jabble) [reply tests/system/python/conftest.py:260]: Okay

**Comment by ashish-jabble** [comment]
@Mohit04tomar I believe it's good idea to use fixture in existing [`tests/system/python/packages/test_rule_data_availability.py`](https://github.com/fledge-iot/fledge/blob/develop/tests/system/python/packages/test_rule_data_availability.py#L92). This way we get tested these changes.

File: tests/system/python/scripts/install_c_plugin [hunk 14]
```diff
@@ -22,11 +22,16 @@ PLUGIN_NAME=$3
...
+if [[ "${PLUGIN_TYPE}" = "rule" || "${PLUGIN_TYPE}" == "notify" ]]; then sudo rm -rf /tmp/fledge-service-notification; fi
 
 
 clean () {
    rm -rf /tmp/${REPO_NAME}
+   if [[ "${PLUGIN_TYPE}" = "rule" ]]; then
+      sudo rm -rf ${FLEDGE_ROOT}/plugins/notificationRule/${PLUGIN_NAME}
+   elif [[ "${PLUGIN_TYPE}" == "notify" ]]; then
+      sudo rm -rf ${FLEDGE_ROOT}/plugins/notificationDelivery/${PLUGIN_NAME}
+   fi
```
Comment (ashish-jabble) [inline tests/system/python/scripts/install_c_plugin:34]: Why sudo is required for notify and rule type?

Comment (ashish-jabble) [reply tests/system/python/scripts/install_c_plugin]: We may use this 
```
clean() {
   rm -rf /tmp/${REPO_NAME}
   if [[ "${PLUGIN_TYPE}" = "rule" ]]; then
      type="notificationRule"
   elif [[ "${PLUGIN_TYPE}" == "notify" ]]; then
      type="notificationDelivery"
   else:
        type=${PLUGIN_TYPE}
   fi
   rm -rf ${FLEDGE_ROOT}/plugins/$type/${PLUGIN_NAME}
}

```

File: tests/system/python/scripts/install_c_plugin [hunk 15]
```diff
@@ -22,11 +22,16 @@ PLUGIN_NAME=$3
 os_name=$(grep -o '^NAME=.*' /etc/os-release | cut -f2 -d\" | sed 's/"//g')
 os_version=$(grep -o '^VERSION_ID=.*' /etc/os-release | cut -f2 -d\" | sed 's/"//g')
 REPO_NAME=fledge-${PLUGIN_TYPE}-${PLUGIN_NAME,,}
-if [[ "${PLUGIN_TYPE}" = "rule" ]]; then rm -rf /tmp/fledge-service-notification; fi
+if [[ "${PLUGIN_TYPE}" = "rule" || "${PLUGIN_TYPE}" == "notify" ]]; then sudo rm -rf /tmp/fledge-service-notification; fi
```
Comment (ashish-jabble) [inline tests/system/python/scripts/install_c_plugin:25]: remove sudo from here as well

File: tests/system/python/conftest.py [hunk 16]
```diff
@@ -212,6 +215,138 @@ def clone_make_install():
 
     return _add_fledge_north
 
+@pytest.fixture
+def add_service():
+    def _add_service(fledge_url, service, service_branch, retries, installation_type = "make", service_name = "play",
+                     use_pip_cache = True, enabled = True):
+        
+        """ 
+            Fixture to add Service and start the start service by default
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:224]: fledge_url is missing

File: tests/system/python/conftest.py [hunk 17]
```diff
@@ -212,6 +215,138 @@ def clone_make_install():
...
+            delivery_config: Configuration of Delivery plugin
+            rule_plugin: Rule plugin to be installed, by default Threshold and DataAvailability plugin is installed 
+            rule_branch: Branch of Rule plugin to be installed
+            rule_plugin_discovery_name: Name to identify the Rule Plugin after installation 
+            delivery_plugin_discovery_name: Name to identify the Delivery Plugin after installation 
+            installation_type: Type of installation for plugins i.e. make or package
+            notification_type: Type of notification to be triggered i.e. one_shot, retriggered, toggle
+            notification_instance_name: Name that will be given to notification instance to be created
+            retrigger_time: Interval between retriggered notifications
+            use_pip_cache: use pip cache in requirements installation
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:289]: only applies with Python based plugins

File: tests/system/python/conftest.py [hunk 18]
```diff
@@ -212,6 +215,138 @@ def clone_make_install():
...
+                     use_pip_cache = True, enabled = True):
+        
+        """ 
+            Fixture to add Service and start the start service by default 
+            service: Service to be installed
+            service_branch: Branch of service to be installed
+            retries: Number of tries for polling
+            installation_type: Type of installation for service i.e. make or package
+            service_name: Name that will be given to service to be installed
+            use_pip_cache: use pip cache in requirements installation
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:230]: only applies with Python based Plugins

File: tests/system/python/conftest.py [hunk 19]
```diff
@@ -212,6 +215,138 @@ def clone_make_install():
...
+        data = {
+                "name": notification_instance_name,
+                "description": "{} notification instance".format(notification_instance_name),
+                "rule_config": rule_config,
+                "rule": rule_plugin_discovery_name,
+                "delivery_config": delivery_config,
+                "channel": delivery_plugin_discovery_name,
+                "notification_type": notification_type,
+                "enabled": enabled, 
+                "retrigger_time": "{}".format(retrigger_time),
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:342]: Can be without format

File: tests/system/python/conftest.py [hunk 20]
```diff
@@ -212,6 +215,138 @@ def clone_make_install():
...
+        assert service_name == retval["name"]
+        return retval
+        
+    return _add_service
+
+@pytest.fixture
+def add_notification_instance():
+    def _add_notification_instance(fledge_url, delivery_plugin, delivery_branch , rule_config={}, delivery_config={}, 
+                                   rule_plugin="Threshold", rule_branch=None, rule_plugin_discovery_name=None, 
+                                   delivery_plugin_discovery_name=None, installation_type='make', notification_type="one shot",
```
Comment (ashish-jabble) [inline tests/system/python/conftest.py:273]: rule_branch None only if rule_plugin exists in Threshold/DataAvailability.
And for rule and delivery plugin discovery names cannot be None in any case.

File: tests/system/python/conftest.py (hunk 19 above)
Comment (Mohit04tomar) [reply tests/system/python/conftest.py:342]: I tried it without format but it configuration is not updated in fledge. That's why I used format.

Comment (ashish-jabble) [reply tests/system/python/conftest.py:342]: I got it as API expects every KV pair in string that's why it is working with format.
Either `str(retrigger_time) or "{}".format(retrigger_time)` both will work.

**Pull request merged**